internal/controller/drplacementcontrol_controller.go

```diff
@@ -166,13 +166,21 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 		// then the DRPC should be deleted as well. The least we should do here is to clean up DPRC.
 		err := r.processDeletion(ctx, drpc, placementObj, logger)
 		if err != nil {
-			logger.Info(fmt.Sprintf("Error in deleting DRPC: (%v)", err))
-
 			statusErr := r.setDeletionStatusAndUpdate(ctx, drpc)
 			if statusErr != nil {
 				err = fmt.Errorf("drpc deletion failed: %w and status update failed: %w", err, statusErr)
+
+				return ctrl.Result{}, err
 			}
 
+			// Is this an expected condition?
+			if rmnutil.IsOperationInProgress(err) {
```

Comment:
Instead of passing back errors, which seems to be of [limited](https://github.com/kubernetes-sigs/controller-runtime/issues/617#issuecomment-589299627) use, and defining an error internally that needs to be honored in other error cases and also existing cases (which adds to code awareness across the functions when returning errors), I would instead just not return error and log it for troubleshooting use.

That way we are not defining an error that other parts of the code needs to return. Current code requests a reconcile using a requeue or an error, which for us is the same. So this change would impact future work.

When starting to write the code initially there was some confusion in this regard, and errors were returned. I suggest we do not return errors, but request a requeue via the result.

Reply:
The current solution is a local change in the in some functions inside the flow, and a local change in Reconcile. This should be easy to backport (and we must backport it to 4.18 since this is a regression added recently).

If we don't use the error, we need to change the all the function signatures in delete flow. We can return `requeue, error`, but this does not allow code to specify a RequeAfter, and even if we don't have code that needs it right now in this flow, we may have code that needs it in other flows or in this flow in the future, so this signature is not future proof. So looks like the new signature should be  `cntl.Result, error`.

But this does not match other controllers, and we probably want a consistent way to return results from controller functions. Do you suggest to refactor entire ramen to return `cntl.Result, error`?

Reply:
For reference, searching for (ctrl.Result, error) show that only few functions use this return value.

Results (without Reconcile functions):

- (r DRClusterReconciler) processCreateOrUpdate
- (r DRClusterReconciler) processDeletion
- (r *DRClusterConfigReconciler) processDeletion
- (r *DRClusterConfigReconciler) processCreateOrUpdate
- (r *DRPlacementControlReconciler) reconcileDRPCInstance
- (r *DRPolicyReconciler) reconcile

Using this signature for deletion flow seems common.

Reply:
This is a well-known and long-standing debate in software design. It's similar to classic discussions around handling conditions like a "file-not-found.", whether you throw an exception or not... Is it an error/exception or simply an expected state? Similarly, in the context of a reconciler, when attempting to fetch a resource, receiving a `NotFound` isn't necessarily a failure but rather an expected condition. But it is still treated as an error.

In our case, the `processDeletion` function returns an error. If it doesn't complete fully, there's clearly a reason for it. This reason holds valuable context that can help the caller logic make informed decisions. Simply logging it without passing it back discards this context.

By returning the error, we make the caller to decide whether to log, requeue, or take any other corrective action. I also believe that `Result{}` should not be propagated from function to function. Only the reconciler function should fill-in that structure with the information returned from the helper functions.

Reply:
I am stating simply that we update result to requeue as `true` on errors and not propogate the Result to other functions. The reconciler is just a loop with an end to a call requesting a requeue or not, the error does not hold much value.

If we add reasons for errors and start making smart decisions we will need to use this pattern consistently, which is a burden, IMHO. If that will not be the case then in this narrow use-case we can do what is done here, but I worry that the pattern will be enforced in the future for no particular gain.

Reply:
> I am stating simply that we update result to requeue as `true` on errors and not propogate the Result to other functions.

So you suggest to change functions in processDeletion flow to behave like this?

```go
func deletteExample(...) bool {
    // Expected to succeed during deletion
    if err := trySomething(...); err != nil {
        log.Error(err, "failed to ...")
        return false
    }

    // Expected to return false for a while during deletion
    if !readyForNextStep(...) {
        return false
    }

    // Completed
    return true
}
```

So we log an INFO message in ensureDeletedVRGs if we the VRG was not deleted yet, and return false.

> If we add reasons for errors and start making smart decisions we will need to use this pattern consistently.

This pattern of trying to do the next step in the way to the desired state is pretty common in ramen. It could be very useful if we use this pattern everywhere. It will make the code much more clear, and avoid all the bogus errors that we have today. This change make it easy to use this pattern everywhere gradually, without changing lot of code.

> If that will not be the case then in this narrow use-case we can do what is done here, but I worry that the pattern will be enforced in the future for no particular gain.

How about using this pattern only in this controller for now? We need a way to fix the regression caused by adding the vrg on the secondary cluster. This should be backported to 4.18 soon before it breaks QE tests in the same way it breaks our tests.

I can also split the fix (the first commit) so we can merge it now. It will cause a lot of noise in the logs during deletion, but we can fix this in the next step when we converge on what is the right way to handle "not ready" state during reconcile.